runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkBatchPortablePipelineTranslator.java

```diff
@@ -224,6 +224,11 @@ private static void translateImpulse(
             MetricsAccumulator.getInstance());
     JavaRDD<RawUnionValue> staged = inputRdd.mapPartitions(function);
 
+    // Prevent potentially expensive re-computation of executable stage
+    if (outputs.size() > 1) {
+      staged.cache();
```

Comment:
For correctness purpose, we should also cache RDDs which have multiple consumers.

Example, ES1 has 1 output PC1 which is consumed by GBKa and GBKb. GBKa is consumed by ESa and GBKb is consumed by ESb. In this case ES1 will be executed twice.

Reply:
Yes, good call. Do you think it would be better to cache all intermediate RDDs (including those for other transforms such as GBK, flatten, etc.), like the legacy Spark runner does, or cache just those outputted by the exec stage function?

Reply:
I went for caching all the intermediate RDDs (can be turned off by `setCacheDisabled(true)`). PTAL

Reply:
I was about to comment the same. 

I think caching all intermediate RDDs will be prohibitively expensive, and disabling them all has re-execution issues. Did you try the flatten + materialize approach we discussed earlier?

Reply:
Flattening the output RDDs (or "leaves") seems to have no effect on re-computation. As far as I understand, Spark caches nothing unless explicitly commanded by the user.

Reply:
That's really unfortunate. 

So the current design is to cache, when the (by default True flag is enabled) (1) collections that are used more than once and (2) all outputs of transforms with multiple outputs.

That seems a reasonable first step. I think we can do better  (e.g. cache all but one output from multi-output transforms, uncache once we're done with all consumers), but could probably call that future work. 

This PR also seems to change a bunch of stuff with how coding works. Could you factor that out into a separate PR?

Reply:
(1) and (2) are correct.

I added the coder as an argument to `SparkTranslationContext::pushDataset` because the coder is used in the Spark runner's bespoke caching implementation: https://github.com/apache/beam/blob/fdedb4c41e32606c64c8c1385d12444998dc4a2b/runners/spark/src/main/java/org/apache/beam/runners/spark/translation/BoundedDataset.java#L110-L117